src/routers/alpha-router/gas-models/mixedRoute/mixed-route-heuristic-gas-model.ts

```diff
@@ -0,0 +1,386 @@
+          },
+          'Debug eth price token issue'
+        );
+        throw err;
+      }
+
+      // true if token0 is the native currency
+      const token0USDPool = usdPool.token0.address == nativeCurrency.address;
+
+      // gets the mid price of the pool in terms of the native token
+      const nativeTokenPriceUSDPool = token0USDPool
+        ? usdPool.token0Price
+        : usdPool.token1Price;
+
+      let gasCostInTermsOfUSD: CurrencyAmount;
+      try {
+        gasCostInTermsOfUSD = nativeTokenPriceUSDPool.quote(
+          totalGasCostNativeCurrency
+        ) as CurrencyAmount;
+      } catch (err) {
+        log.info(
+          {
+            usdT1: usdPool.token0.symbol,
+            usdT2: usdPool.token1.symbol,
+            gasCostInNativeToken: totalGasCostNativeCurrency.currency.symbol,
+          },
+          'Failed to compute USD gas price'
+        );
+        throw err;
+      }
+
+      return {
+        gasEstimate: baseGasUse,
+        gasCostInToken: gasCostInTermsOfQuoteToken,
+        gasCostInUSD: gasCostInTermsOfUSD!,
+      };
+    };
+
+    return {
+      estimateGasCost: estimateGasCost.bind(this),
+    };
+  }
+
+  private estimateGas(
+    routeWithValidQuote: MixedRouteWithValidQuote,
+    gasPriceWei: BigNumber,
+    chainId: ChainId
+  ) {
+    const totalInitializedTicksCrossed = BigNumber.from(
+      Math.max(1, _.sum(routeWithValidQuote.initializedTicksCrossedList))
+    );
+    /**
+     * Since we must make a separate call to multicall for each v3 and v2 section, we will have to
+     * add the BASE_SWAP_COST to each section.
+     */
+    let baseGasUse = BigNumber.from(0);
+
+    const route = routeWithValidQuote.route;
+    let acc = [];
+    let j = 0;
+    while (j < route.pools.length) {
+      // seek forward until finding a pool of different type
+      let section = [];
+      if (route.pools[j] instanceof Pool) {
+        while (route.pools[j] instanceof Pool) {
+          section.push(route.pools[j]);
+          j++;
+          if (j === route.pools.length) {
+            // we've reached the end of the route
+            break;
+          }
+        }
+        acc.push(section);
+        /// we just added a complete v3 section
+        baseGasUse = baseGasUse.add(BASE_SWAP_COST(chainId));
+        baseGasUse = baseGasUse.add(COST_PER_HOP(chainId).mul(section.length));
+      } else {
+        while (route.pools[j] instanceof Pair) {
+          section.push(route.pools[j]);
+          j++;
+          if (j === route.pools.length) {
+            // we've reached the end of the route
+            break;
+          }
+        }
+        acc.push(section);
+        /// we just added a complete v2 section
+        baseGasUse = baseGasUse.add(BASE_SWAP_COST_V2);
+        baseGasUse = baseGasUse.add(
+          /// same behavior in v2 heuristic gas model factory
+          COST_PER_EXTRA_HOP_V2.mul(section.length - 1)
+        );
+      }
+    }
```

Comment:
This extra logic for dealing with v2 gas in addition to v3 stuff is the only real difference (besides typing) from the v3 gas model. Wonder if it's worth exploring a refactor.